### PR TITLE
num/dualquat: package for dual quaternion maths

### PR DESCRIPTION
**Pull request by kortschak**
I have struggled with getting the trig functions to work with these packages and this has prevented me from sending this. Given that the primary use for these types of numbers is to allow 2D and 3D translations and rotations, something that doesn't require trig in this number space, I have decided that discretion is the better part of valor and have removed them. The problems I was facing were compounded (but not entirely caused) by the odd behaviour of "math/cmplx" for some of these functions. If people want trig functions for these, they can file an issue and then I can look at it again (the almost working code is in dualquat-full).

~There is one issue that I am not entirely sure of yet, and I think stems from the commutativity of complex numbers when quaternions are not commutative (this shows up in the different behaviours of the rotate and translate examples of the two packages). I would like to sort this out, but am sending this now.~

Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
I've spent a bit of time messing about trying to get the dual complex examples to work, and I know what it not working, but don't understand how to fix that. The issue is that rotations leave the non-dual as non-unit which they should not. However, using the approach used for quaternion rotation (or dual quaternion rotation/translation does not work).

As a note for myself, read [this](https://arxiv.org/abs/1601.01754v1).

**Comment by kortschak** [comment]
I have backed out the dual complex until I understand it better. It will be a separate PR.

File: num/dualquat/dual.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+		fre := fmtString(fs, c, prec, width, false)
+		fim := fmtString(fs, c, prec, width, true)
+		fmt.Fprintf(fs, fmt.Sprintf("(%s+%[2]sϵ)", fre, fim), d.Real, d.Dual)
+	default:
+		fmt.Fprintf(fs, "%%!%c(%T=%[2]v)", c, d)
+		return
+	}
+}
+
+// This is horrible, but it's what we have.
```
Comment (btracey) [inline num/dualquat/dual.go]: Almost not though!

File: num/dualquat/dual.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+// ConjDual returns the dual conjugate of d₁+d₂ϵ, d₁-d₂ϵ.
+func ConjDual(d Number) Number {
+	return Number{
+		Real: d.Real,
+		Dual: quat.Scale(-1, d.Dual),
+	}
+}
+
+// ConjQuat returns the quaternion conjugate of d₁+d₂ϵ, d̅₁+d̅₂ϵ.
```
Comment (btracey) [inline num/dualquat/dual.go]: Is there a need to have a `ConjDualQuat` that does both?

File: num/dualquat/dual_example_test.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	return dualquat.Number{
+		Real: quat.Number{Real: 1},
+		Dual: raise(p),
+	}
+}
+
+// transform performs the quaternion rotation of p by the given quaternion
+// and scaling by the scale factor. The rotations are normalized to unit
+// vectors.
+func transform(p point, by ...dualquat.Number) point {
```
Comment (btracey) [inline num/dualquat/dual_example_test.go]: Is this something that should exported? It's the core of the Example. Or is the idea that some outside package would do this?

File: num/dualquat/dual_example_test.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+		q = dualquat.Mul(o, q)
+	}
+	pp := dualquat.Mul(dualquat.Mul(q, raiseDual(p)), dualquat.ConjDual(dualquat.ConjQuat(q)))
+
+	// Extract the point.
+	return point{x: pp.Dual.Imag, y: pp.Dual.Jmag, z: pp.Dual.Kmag}
+}
+
+func Example() {
+	// Translate a 1×1×1 cube [3, 4, 5] and rotate it 120° around the
```
Comment (btracey) [inline num/dualquat/dual_example_test.go]: s/cube [3,/cube by [3,

File: num/dualquat/dual_example_test.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+		// Clean up floating point error for clarity.
+		pp.x = floats.Round(pp.x, 2)
+		pp.y = floats.Round(pp.y, 2)
+		pp.z = floats.Round(pp.z, 2)
+
+		fmt.Printf(" %d %+v -> %+v\n", i, p, pp)
+	}
+
+	// Rotate a line segment from [2, 1, 1] to [2, 1, 2] 120° around
```
Comment (btracey) [inline num/dualquat/dual_example_test.go]: Maybe "Rotate the line segment {[2, 1, 1], [2, 1, 2]}"

File: num/dualquat/dual_example_test.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+	// Construct a displacement.
+	displace := dualquat.Number{
+		Real: quat.Number{Real: 1},
+		Dual: quat.Scale(0.5, raise(point{3, 4, 5})),
+	}
+
+	// Construct a rotations.
+	alpha := 2 * math.Pi / 3
+	axis := raise(point{1, 1, 1})
```
Comment (btracey) [inline num/dualquat/dual_example_test.go]: You should add a comment for what it does, as it won't be in the godoc.

File: num/dualquat/dual_fike.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Derived from code by Jeffrey A. Fike at http://adl.stanford.edu/hyperdual/
+
+// The MIT License (MIT)
```
Comment (btracey) [inline num/dualquat/dual_fike.go]: Why does this need a license? It seems like this code is yours using the other code that is related to Jeff's

File: num/dualquat/dual_simple_example_test.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+package dualquat_test
+
+import (
+	"fmt"
+
+	"gonum.org/v1/gonum/num/dualquat"
+	"gonum.org/v1/gonum/num/quat"
+)
+
+// Example point, displacement and rotation from Euclidean Space Dual Quaternions page:
```
Comment (btracey) [inline num/dualquat/dual_simple_example_test.go]: It seems that this is worth including in the example text, the website seems nice.

File: num/dualquat/doc.go [hunk 9]
```diff
@@ -0,0 +1,9 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package dualquat provides the dual quaternion numeric type and functions.
```
[the remark this answers is not in the document]
Comment (kortschak) [reply num/dualquat/doc.go]: Added.

I can add a similar what to dual and hyperdual if you think it's worth it here.

File: num/dualquat/dual.go (hunk 1 above)
Comment (kortschak) [reply num/dualquat/dual.go]: ?

File: num/dualquat/dual.go (hunk 2 above)
Comment (kortschak) [reply num/dualquat/dual.go]: Yeah. I was thinking about that. For the dual quaternion case, there are descriptions in the literature of both the dual conjugate and the quaternion conjugate, but only the double is used for this case. In the dual complex system, only the one conjugate is described as far as I can see. I am thinking that I could add just `Conj` which does the obviously correct thing and then leave the others as helpers, or even just remove them.

I'll add `Conj` and leave the others for now.

File: num/dualquat/dual.go (hunk 1 above)
Comment (btracey) [reply num/dualquat/dual.go]: 1.12 is almost out, which means 1.9 support will be dropped from Gonum

File: num/dualquat/dual_example_test.go (hunk 3 above)
Comment (kortschak) [reply num/dualquat/dual_example_test.go]: This is factored out to help with the explanation. I wouldn't do it this way.

File: num/dualquat/dual_example_test.go (hunk 4 above)
Comment (kortschak) [reply num/dualquat/dual_example_test.go]: Done

File: num/dualquat/dual_example_test.go (hunk 5 above)
Comment (kortschak) [reply num/dualquat/dual_example_test.go]: Done

File: num/dualquat/dual_example_test.go (hunk 6 above)
Comment (kortschak) [reply num/dualquat/dual_example_test.go]: No, this is in the godoc. See the "Larger Examples" section in https://blog.golang.org/examples.

File: num/dualquat/dual_fike.go (hunk 7 above)
Comment (kortschak) [reply num/dualquat/dual_fike.go]: The is very similar to the code from Jeff's for the dual reals. The dual complex is adequately different and I removed the distinction there. I don't mind either way, but if we do remove it, I'd prefer to check with him first. You know him, so do you want to ask?

File: num/dualquat/dual_simple_example_test.go (hunk 8 above)
Comment (kortschak) [reply num/dualquat/dual_simple_example_test.go]: Added in the package doc.

File: num/dualquat/dual.go (hunk 1 above)
Comment (kortschak) [reply num/dualquat/dual.go]: Ah. That's not the horrible thing. The horrible thing is that the `fmt.Formatter` API does not allow a pass through without constructing a new format string. I can just say, format these things with the same format as I already have, so we have to perform at least one more allocation and then reparse the format string.

**Comment by kortschak** [comment]
References are added to the package doc.

Also https://dualquaternions.weebly.com/uploads/1/3/1/5/13150194/dual_quaternions_wscg_paper_pub.pdf

I just remembered the WP page is OK, https://en.wikipedia.org/wiki/Dual_quaternion. I'll add that to the package doc.

File: num/dualquat/dual_example_test.go (hunk 6 above)
Comment (btracey) [reply num/dualquat/dual_example_test.go]: Thanks, I didn't realize that.

File: num/dualquat/dual_fike.go (hunk 7 above)
Comment (btracey) [reply num/dualquat/dual_fike.go]: It's okay to leave in, I looked at the other code and it's more similar than I remembered.

File: num/dualquat/dual.go (hunk 2 above)
Comment (btracey) [reply num/dualquat/dual.go]: SGTM

**Pull request merged**